lighthouse/aggregate-service-import.md

```diff
@@ -0,0 +1,137 @@
+# Aggregate Service Import for Lighthouse
+
+## Summary
+
+The current Lighthouse implementation deviates from the MCS specification in the following ways:
+
+- ServiceExports with matching namespace/name are not combined into a single ServiceImport.
+  Each importing cluster receives multiple ServiceImports that are multiplexed at DNS lookup.
+- Conflict resolution of ServiceExports is not implemented [^1].
+
+In order to better align with the MCS specification the following functionality needs to be
+added to the Lighthouse agent:
+
+- Merge matching `ServiceExport` objects into a single `ServiceImport` object
+- Mark conflicting `ServiceExport` objects with the `ServiceExportConflict` condition
+
+There are changes required in Lighthouse's handling of `EndpointSlices` before single
+`ServiceImports` can be introduced.
+
+- The `ServiceImports` for an exported service each contain the service VIP from the exporting
+  cluster. `ServiceImports` cannot be merged like this because a `ServiceImport` cannot have
+  more than one IP. The semantics of the IP are that it should be a cluster local Service VIP,
+  if needed. The service VIPs from the exporting clusters should be stored in `EndpointSlices`,
+  with a separate `EndpointSlice` per exporting cluster.
+
+It is worth noting that:
+
+- Headless services already have `EndpointSlices` containing the correct pod IP addresses, and contain
+  no VIP information in the multiple `ServiceImports`. A merged  `ServiceImport` for headless
+  services would comprise a merge of the port definitions in those `ServiceImports`.
+
+- The current GlobalNet implementation exports `EndpointSlices` for normal services that contain
+  pod IP addresses from the originating cluster.
+
+## Proposal
+
+Submariner has a passive broker, it is a central apiserver that is accessed by each cluster in
+the clusterset. Clusters synchronise resources to and from the broker with all sync and
+reconciliation logic implemented in the clusters.
+
+In this proposal we retain the passive broker and take an approach to merging `ServiceImports`
+where exporting clusters handle the merging. Simplistically, a merged `ServiceImport` outcome
+can be achieved by allowing the first exporting cluster to create a `ServiceImport` and all
+subsequent exporting clusters to reuse it, assuming they have matching port lists.
+
+1. First exporting cluster creates a `ServiceImport` on the broker.
+1. Subsequent exporting clusters reuse the existing `ServiceImport`.
+1. Exporting clusters create and sync `EndpointSlices` to the broker with the service VIP and a
+   `multicluster.kubernetes.io/service-name` label that matches the `ServiceImport`.
```

Comment:
I think we can also think of maintaining ServiceImport per cluster which contain the cluster-IP of the exported cluster in the submariner-operator namespace of each cluster and have an aggregated service import in the Service Namespace. Conflict resolution still can happen based on the presence of a ServiceImport in the submariner-operator namespace.

This way we can leave the endpointslices as it is. Though Submariner does not create multiple endpointslices now it is expected that an enpointslice should contain only a specified number of endpoints (100 by default). So there may be multiple endpoint slices per service per cluster.

Reply:
`EndpointSlices` can remain as they are for non-global net deployments, but this implies not using service load balancing in the destination cluster, which I thought was desired for normal clusterip services. For globalnet they currently contain unreachable pod IPs for normal clusterip services. I think it is desirable to change `EndpointSlice` to contain the target cluster's globalnet service VIP for clusterip services and retain the existing behaviour for headless services.